src/System.Reflection.Metadata/src/System/Reflection/PortablePdb/PortablePdbReader.cs

```diff
@@ -0,0 +1,213 @@
+        /// 
+        /// Unless <see cref="PdbStreamOptions.LeaveOpen"/> is specified, ownership of the stream is transferred to the <see cref="PortablePdbReader"/> 
+        /// upon successful argument validation. It will be disposed by the <see cref="PortablePdbReader"/> and the caller must not manipulate it.
+        /// 
+        /// Unless <see cref="PdbStreamOptions.PrefetchMetadata"/> is specified no data 
+        /// is read from the stream during the construction of the <see cref="PortablePdbReader"/>. Furthermore, the stream must not be manipulated
+        /// by caller while the <see cref="PortablePdbReader"/> is alive and undisposed.
+        /// 
+        /// If <see cref="PdbStreamOptions.PrefetchMetadata"/>, the <see cref="PortablePdbReader"/> 
+        /// will have read all of the data requested during construction. As such, if <see cref="PdbStreamOptions.LeaveOpen"/> is also
+        /// specified, the caller retains full ownership of the stream and is assured that it will not be manipulated by the <see cref="PortablePdbReader"/>
+        /// after construction.
+        /// </param>
+        /// <exception cref="ArgumentOutOfRangeException">Size is negative or extends past the end of the stream.</exception>
+        public PortablePdbReader(Stream pdbStream, PdbStreamOptions options, int size)
+            : this(pdbStream, options, (int?)size)
+        {
+        }
+
+        private unsafe PortablePdbReader(Stream pdbStream, PdbStreamOptions options, int? sizeOpt)
+        {
+            if (pdbStream == null)
+            {
+                throw new ArgumentNullException(nameof(pdbStream));
+            }
+
+            if (!pdbStream.CanRead || !pdbStream.CanSeek)
+            {
+                throw new ArgumentException(SR.StreamMustSupportReadAndSeek, nameof(pdbStream));
+            }
+
+            if (!options.IsValid())
+            {
+                throw new ArgumentOutOfRangeException(nameof(options));
+            }
+
+            long start = pdbStream.Position;
+            int size = StreamExtensions.GetAndValidateSize(pdbStream, sizeOpt, nameof(pdbStream));
+
+            bool closeStream = true;
+            try
+            {
+                bool isFileStream = FileStreamReadLightUp.IsFileStream(pdbStream);
+
+                if ((options & PdbStreamOptions.PrefetchMetadata) == 0)
+                {
+                    _lazyMetadataBlock = null;
+                    _provider = new StreamMemoryBlockProvider(pdbStream, start, size, isFileStream, (options & PdbStreamOptions.LeaveOpen) != 0);
+                    closeStream = false;
+                }
+                else
+                {
+                    // Read in the entire image or metadata blob:
+
+                    _lazyMetadataBlock = StreamMemoryBlockProvider.ReadMemoryBlockNoLock(pdbStream, isFileStream, 0, (int)Math.Min(pdbStream.Length, int.MaxValue));
```

Comment:
I think we have a bug here carried over from PEREader. :( Should that 0 be start? Looks like we always read from the start of stream when prefetch is passed but not otherwise. Docs say we read from current position.